File: src/main/java/org/moinex/model/investment/MarketQuotesAndCommodities.java
/*
 * Filename: MarketQuotesAndCommodities.java
 * Created on: January 17, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.model.investment;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "market_quotes_and_commodities")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class MarketQuotesAndCommodities
{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Long id;

    @Column(name = "dollar")
    private BigDecimal dollar;

    @Column(name = "euro")
    private BigDecimal euro;

    @Column(name = "ibovespa")
    private BigDecimal ibovespa;

    @Column(name = "bitcoin")
    private BigDecimal bitcoin;

    @Column(name = "ethereum")
    private BigDecimal ethereum;

    @Column(name = "gold")
    private BigDecimal gold;

    @Column(name = "soybean")
    private BigDecimal soybean;

    @Column(name = "coffee")
    private BigDecimal coffee;

    @Column(name = "wheat")
    private BigDecimal wheat;

    @Column(name = "oil_brent")
    private BigDecimal oilBrent;

    @Column(name = "last_update")
    private LocalDateTime lastUpdate;
}

File: src/main/kotlin/org/moinex/model/investment/MarketQuotesAndCommodities.kt
/*
 * Filename: MarketQuotesAndCommodities.kt (original filename: MarketQuotesAndCommodities.java)
 * Created on: January 17, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.model.investment

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.moinex.common.extension.toRounded
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "market_quotes_and_commodities")
class MarketQuotesAndCommodities(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @Column(name = "dollar")
    var dollar: BigDecimal? = null,
    @Column(name = "euro")
    var euro: BigDecimal? = null,
    @Column(name = "ibovespa")
    var ibovespa: BigDecimal? = null,
    @Column(name = "bitcoin")
    var bitcoin: BigDecimal? = null,
    @Column(name = "ethereum")
    var ethereum: BigDecimal? = null,
    @Column(name = "gold")
    var gold: BigDecimal? = null,
    @Column(name = "soybean")
    var soybean: BigDecimal? = null,
    @Column(name = "coffee")
    var coffee: BigDecimal? = null,
    @Column(name = "wheat")
    var wheat: BigDecimal? = null,
    @Column(name = "oil_brent")
    var oilBrent: BigDecimal? = null,
    @Column(name = "last_update")
    var lastUpdate: LocalDateTime? = null,
) {
    init {
        dollar = dollar?.toRounded()
        euro = euro?.toRounded()
        ibovespa = ibovespa?.toRounded()
        bitcoin = bitcoin?.toRounded()
        ethereum = ethereum?.toRounded()
        gold = gold?.toRounded()
        soybean = soybean?.toRounded()
        coffee = coffee?.toRounded()
        wheat = wheat?.toRounded()
        oilBrent = oilBrent?.toRounded()
    }

    override fun toString(): String = "MarketQuotesAndCommodities [id=$id, lastUpdate=$lastUpdate]"
}
